IQSS/dataverse-frontend
# Dynamically render form field in Create Dataset Form + Fields Validations

## What this PR does / why we need it:
Dynamically renders all fields of each metadatablock in the dataset creation form and adds their respective validations.

This sub branch is a combination of two issues that are tightly related.
 - https://github.com/IQSS/dataverse-frontend/issues/336
 - https://github.com/IQSS/dataverse-frontend/issues/338

Both issues have their own pr's merged to this sub branch:
- https://github.com/IQSS/dataverse-frontend/pull/357
- https://github.com/IQSS/dataverse-frontend/pull/368

## Which issue(s) this PR closes:

- Closes 
 https://github.com/IQSS/dataverse-frontend/issues/336
 https://github.com/IQSS/dataverse-frontend/issues/338

## Special notes for your reviewer:
[Create Dataset Form Storybook](https://646f68aa9beb01b35c599acd-nvwtdevnnw.chromatic.com/?path=/story/pages-create-dataset--default)

## Suggestions on how to test this:
### Step 1: Run the Development Environment
1. Execute `npm i`.
2. Navigate with `cd packages/design-system && npm run build`.
3. Return with `cd ../../`.
4. Ensure you have a `.env` file similar to `.env.example`, with the variable `VITE_DATAVERSE_BACKEND_URL=http://localhost:8000`.
5. Navigate with `cd dev-env`.
6. Start the environment using `./run-env.sh unstable`.
7. To verify the environment, visit <http://localhost:8000> and check your local Dataverse installation.

### Step 2: Test the Create Dataset Form
- Log in, click the create dataset button and the form fields should be displayed successfully.
- Submit the form without filling any required field, every required field should show te required error feedback, and the form should focus and scroll to the first input with an error.
- Fill a date field with the wrong format, respective feedback error should be visible.
- Fill an integer field with the wrong format (e.g.: 23.5), respective feedback error should be visible.
- Fill a float field with the wrong format (e.g.: 'hi'), respective feedback error should be visible.
- Fill an email field with a wrong format (e.g.: 'test@'), respective feedback error should be visible.
- Fill a URL field with a wrong format (e.g.: 'bad://something'), respective feedback error should be visible.
- Submit the form filling al required fields with the correct format, a new dataset should be created.

Run unit and e2e tests and check they are all passing.

## Does this PR introduce a user interface change? If mockups are available, please link/include them here:
Yes.
![spa_datasets_create(Desktop 1366)](https://github.com/IQSS/dataverse-frontend/assets/64560524/307892e1-e8ce-406b-900a-1f239f7700b6)

Also I modified at spa level not design system level the styles for the checkbox group, I think if it is successful we could apply it to the design system directly in the future. 
This checkbox group that allows to select multiple values is a temporary replacement of the [select multiple](https://github.com/IQSS/dataverse-frontend/issues/355) not yet done.
Examples below: 👇👇
![Screenshot 2024-04-11 at 14 01 29](https://github.com/IQSS/dataverse-frontend/assets/64560524/1792e9d7-87d3-451b-9b13-81b2b350c412)
![Screenshot 2024-04-11 at 14 01 42](https://github.com/IQSS/dataverse-frontend/assets/64560524/2f63d024-3053-415a-ba93-10123b403732)
![Screenshot 2024-04-11 at 14 01 51](https://github.com/IQSS/dataverse-frontend/assets/64560524/e4ca90a5-ac92-45fb-9175-1626671858fc)

## Is there a release notes update needed for this change?:

## Additional documentation:
I have refactor some design system components:
- Accordion: refactor of props using already existen props from react-bootstrap.
- AccordionBody: able to receive `bsPrefix` and `as` props and extend html props attributes.
- AccordionHeader: able to receive `onClick`, `bsPrefix` and `as` props and extend html props attributes.
- AccordionItem: able to receive `bsPrefix` and `as` props and extend html props attributes.
- FormGroup: able to clone children but wrapped through fragments also.
- Some form elements so that they can forward a react ref and pass it to the input itself. This was necessary so that the form validation library could focus and scroll to the first item encountered with a validation error.

All tests were ok and add one more test to test the cloning through childrens wrapped in fragments.
